lib/Sema/CSSimplify.cpp

```diff
@@ -4675,6 +4675,12 @@ ConstraintSystem::matchTypesBindTypeVar(
                : getTypeMatchFailure(locator);
   }
 
+  if (typeVar->getImpl().isKeyPathType()) {
+    return resolveKeyPath(typeVar, type, flags, locator)
```

Comment:
Doing this in matchTypesBindTypeVar() seems wrong. Why can't you generate these constraints when the type variable is created?

Reply:
The key paths are special in a way that we need to infer capability from the members in the chain before we can form the type.

Reply:
To elaborate more, when type variable is created there is no type for it. The type we get here is purely synthetic, formed by inference based on capabilities inferred from members referenced in a key path, which means that the type wasn't previously "opened" like in other places and has to be opened before it's assigned. I agree that this is unfortunate and we'd need a better way to express this but I don't see how we can do that without significant changes to inference.